cpu/cc2538/lpm.c

```diff
@@ -212,13 +222,40 @@ lpm_exit()
    * We need to convert sleep duration from rtimer ticks to clock ticks */
   clock_adjust();
 
-  /* Restore system clock to the 32 MHz XOSC */
+  /* CC2538 silicon errata:
+     Possible Incorrect Value of Clock Dividers after PM2 and PM3 */
+#if SYS_CTRL_SYS_DIV == SYS_CTRL_CLOCK_CTRL_SYS_DIV_32MHZ || \
+  SYS_CTRL_IO_DIV == SYS_CTRL_CLOCK_CTRL_IO_DIV_32MHZ
+  if(reset_div_mask) {
+    /* We come from PM2 or PM3 where we adjusted (=set to 16 MHz)
+       either one or both divisors according to the errata note.
+       Set them back to 32 MHz. */
+    uint32_t val = REG(SYS_CTRL_CLOCK_CTRL);
+
+#if SYS_CTRL_SYS_DIV == SYS_CTRL_CLOCK_CTRL_SYS_DIV_32MHZ && \
+    SYS_CTRL_IO_DIV == SYS_CTRL_CLOCK_CTRL_IO_DIV_32MHZ
+    /* Set both divisors back to 0 (=32MHz) */
+    val &= ~(SYS_CTRL_CLOCK_CTRL_SYS_DIV + SYS_CTRL_CLOCK_CTRL_IO_DIV);
+#elif SYS_CTRL_SYS_DIV == SYS_CTRL_CLOCK_CTRL_SYS_DIV_32MHZ
+    /* Set sys clock divisor back to 0 (=32MHz) */
+    val &= ~SYS_CTRL_CLOCK_CTRL_SYS_DIV;
+#elif SYS_CTRL_IO_DIV == SYS_CTRL_CLOCK_CTRL_IO_DIV_32MHZ
+    /* Set io clock divisor back to 0 (=32MHz) */
+    val &= ~SYS_CTRL_CLOCK_CTRL_IO_DIV;
+#endif
+
+    REG(SYS_CTRL_CLOCK_CTRL) = val;
+    reset_div = 0;
+  }
+#endif
```

Comment:
`select_16_mhz_rcosc()` can no longer work properly with the dividers set for 32 MHz. The required change to fix this is also the first step of the workaround for the erratum, so the restore stage has to be in `select_32_mhz_xosc()`, which makes things very easy and avoids duplicating the restore code like you did.

Moreover, this code can be simplified a lot.

Just replace the first line of `select_16_mhz_rcosc()` with this:
```
  REG(SYS_CTRL_CLOCK_CTRL) = (REG(SYS_CTRL_CLOCK_CTRL)
#if SYS_CTRL_SYS_DIV == SYS_CTRL_CLOCK_CTRL_SYS_DIV_32MHZ
    | SYS_CTRL_CLOCK_CTRL_SYS_DIV_16MHZ
#endif
#if SYS_CTRL_IO_DIV == SYS_CTRL_CLOCK_CTRL_IO_DIV_32MHZ
    | SYS_CTRL_CLOCK_CTRL_IO_DIV_16MHZ
#endif
    ) & ~SYS_CTRL_CLOCK_CTRL_OSC_PD;
```

And replace the last line of `select_32_mhz_xosc()` with this:
```
  REG(SYS_CTRL_CLOCK_CTRL) = (REG(SYS_CTRL_CLOCK_CTRL)
#if SYS_CTRL_SYS_DIV == SYS_CTRL_CLOCK_CTRL_SYS_DIV_32MHZ
    & ~SYS_CTRL_CLOCK_CTRL_SYS_DIV
#endif
#if SYS_CTRL_IO_DIV == SYS_CTRL_CLOCK_CTRL_IO_DIV_32MHZ
    & ~SYS_CTRL_CLOCK_CTRL_IO_DIV
#endif
    ) | SYS_CTRL_CLOCK_CTRL_OSC_PD;
```